File: Lista01/Java/ex05.java
import java.util.ArrayList;
import java.util.List;

public class ex05 {
    // Classe base Animal
    public static class Animal {
        void som() {
            System.out.println("O animal faz um som.");
        }
    }

    // Classe Cachorro que herda de Animal
    public static class Cachorro extends Animal {
        @Override
        void som() {
            System.out.println("O cachorro faz: Au Au!");
        }
    }

    // Classe Gato que herda de Animal
    public static class Gato extends Animal {
        @Override
        void som() {
            System.out.println("O gato faz: Miau!");
        }
    }

    // Método que recebe uma lista de objetos Animal
    public static void emitirSons(List<Animal> animais) {
        for (Animal animal : animais) {
            animal.som();
        }
    }

    public static void main(String[] args) {
        List<Animal> animais = new ArrayList<>();
        animais.add(new Cachorro());
        animais.add(new Gato());

        emitirSons(animais);
    }
}

File: Lista01/PY/ex05.py
# Classe base Animal
class Animal:
    def __init__(self, nome):
        self.nome = nome

    # Método genérico de som (a ser sobrescrito nas subclasses)
    def som(self):
        pass

# Classe derivada Cachorro
class Cachorro(Animal):
    def som(self):
        print(f"{self.nome} faz: Au Au Au!")

# Classe derivada Gato
class Gato(Animal):
    def som(self):
        print(f"{self.nome} faz: Miauuu!")

# Classe derivada Vaca (para adicionar mais um exemplo)
class Vaca(Animal):
    def som(self):
        print(f"{self.nome} faz: Muuuu!")

# Função que recebe uma lista de objetos Animal e chama o método som
def emitir_sons(animais):
    for animal in animais:
        animal.som()

# Exemplo de uso
cachorro = Cachorro("Pretinho")
gato = Gato("Olívio")
vaca = Vaca("Joa")

# Criando uma lista de objetos de diferentes classes
lista_animais = [cachorro, gato, vaca]

# Chamando a função que emite os sons de cada animal
emitir_sons(lista_animais)
